K3po nukleus extension tests

NicoletaOita opened the pull request:
Added k3po nukleus tests for window and throttle options for each of the transmissions modes (simplex, half-duplex, duplex).

dpwspoon commented on src/test/java/org/reaktivity/k3po/nukleus/ext/DuplexIT.java:
```diff
@@ -318,6 +318,65 @@ public void shouldThrottleInitialOnlyServerSentData() throws Exception
+    @Ignore
+    // should succeed because the receiver window is 0 and window adjustment is enabled
+    @Test
+    @Specification({
```
I think if the window is set to 0 then it will be 0 forever... I would think?

dpwspoon commented on src/test/java/org/reaktivity/k3po/nukleus/ext/HalfDuplexIT.java:
```diff
@@ -338,6 +338,65 @@ public void shouldOverflowServerSentData() throws Exception
+    // should succeed because the receiver window is 0 and window adjustment is enabled
+    @Test
+    @Specification({
+        "server.set.window0.throttle.on/client",
```
I would expect this to fail for the same reason as above

dpwspoon commented on src/test/java/org/reaktivity/k3po/nukleus/ext/SimplexIT.java:
```diff
@@ -178,6 +189,65 @@ public void shouldThrottleInitialOnlyClientSentData() throws Exception
+    // should succeed because the receiver window is 0 and window adjustment is enabled
+    @Test
+    @Specification({
+        "server.set.window0.throttle.on/client",
```
I would expect this to fail for the same reason as above

dpwspoon commented:
These look good to me minus the confusion on the one test.  Please change that to pass (I think).  @jfallows do you confirm?

NicoletaOita commented:
@dpwspoon I've updated the tests as requested and added a note in the readme related to the window and throttle options. Please merge if it's ok.

cmebarrow commented:
@NicoletaOita I don't understand how any tests in k3po nukleus extension can be related to TCP. Please change the title of this PR to state what it is really doing and add a description if needed.

NicoletaOita commented:
@cmebarrow Indeed, these are not tcp tests, I've started to look over them as part of the tcp story, but they are k3po-nukleus-extension tests. I've updated the name and the description of the story.

cmebarrow commented on README.md:
```diff
@@ -213,3 +213,8 @@ read aborted
+
+NOTE: 
+The option window refers to the size/number of bytes the receiver can accept. 
+The option throttle refers to the way the receiver adjusts its window (when set to "message" or "stream" it's on, when set to "none" it's off). Currently when window is set to 0, it will remain 0 even if throttle is set to "on".
```
I don't find this very clear. Here's a proposal for rewording:

NOTES:
- The "nukleus:window" option specifies the number of bytes that will be advertised in the first Window frame sent back in response to a Begin
- The "nukleus:throttle" option determines how and whether writes will respect Window frames received from the other end. When set to "stream" or "message", window frames are honored, When set to "none" they are not.

NicoletaOita replied:
Ok, I've rewritten it.

cmebarrow commented on src/test/java/org/reaktivity/k3po/nukleus/ext/DuplexIT.java:
```diff
@@ -318,6 +318,66 @@ public void shouldThrottleInitialOnlyServerSentData() throws Exception
+    @Test
+    @Specification({
+        "client.sent.throttle.off/client",
+        "client.sent.throttle.off/server"
```
This test will not pass reliably because there is a race between the client sending the first Data frame and the server sending the initial Window frame. If the server receives the Data frame before it writes the Window frame the test will fail because the server will (correctly) reply with a Reset frame. Moreover, there is already a test that shows that receiving a message that fits in the initial window works as expected: client.sent.throttle.initial.only. So please remove this test.

NicoletaOita replied:
Indeed, I haven't thought about the race, the test passed for me always, but I guess I was just lucky. I've removed it.

cmebarrow commented on src/test/java/org/reaktivity/k3po/nukleus/ext/DuplexIT.java:
```diff
@@ -318,6 +318,66 @@ public void shouldThrottleInitialOnlyServerSentData() throws Exception
+        "client.sent.throttle.off.negative/client",
+        "client.sent.throttle.off.negative/server"
+    })
+    public void shouldNotReceiveMessageThrottleOff() throws Exception
```
If I understand correctly the purpose of this test is to show (a) the client sends data immediately without waiting for a window frame or honoring it (b) the receiver correctly rejects the data because the window has been violated. The rejection is done by sending a Reset frame.
Your test should be testing directly for this rather than just expecting a general test failure. We should build a do this by adding "write aborted" to the client script at the end (after write "Hello, World").
After figuring this out, I looked for a script that is already doing this and found it: server.sent.overflow.

So this test is a duplicate and should be removed. BTW, it would be good to fix the server.sent.overflow/client.rpt script to set nukleus:throttle to "none" instead of "off" (which is now deprecated).

NicoletaOita replied:
Ok, I've removed it and also fixed the throttle option to "none" everywhere it was used.

cmebarrow commented on src/test/java/org/reaktivity/k3po/nukleus/ext/DuplexIT.java:
```diff
@@ -318,6 +318,66 @@ public void shouldThrottleInitialOnlyServerSentData() throws Exception
+        "server.set.window0.throttle.off/client",
+        "server.set.window0.throttle.off/server"
+    })
+    public void shouldNotReceiveMessageWindow0ThrottleingOff() throws Exception
```
Like shouldNotReceiveMessageThrottleOff this test is a duplicate of server.sent.overflow and should be removed. There's nothing special about setting window to 0, it's just another value, There no more reason to test it than to test say value 756. If we had defined special behavior for window 0 (like not sending a window frame for that case) then it would be worth testing, but we have not, it just causes us to send a window frame with 0 bytes.

NicoletaOita replied:
Allright, removed it.

cmebarrow commented on src/test/java/org/reaktivity/k3po/nukleus/ext/DuplexIT.java:
```diff
@@ -318,6 +318,66 @@ public void shouldThrottleInitialOnlyServerSentData() throws Exception
+        "server.set.window0.throttle.on/client",
+        "server.set.window0.throttle.on/server"
+    })
+    public void shouldReceiveMessageWindow0ThrottleingOn() throws Exception
```
I think this test will fail due to a timeout. That means it will take 10 seconds, and that's multiplied by three including the other tests (SimplexIT and HalfDuplexIT) so will significantly increase the build time. The purpose of the test I think is to verify that the writer does not violate the window (i.e. write data larger than the window). That is already amply verified by most of the other tests: they would all fail if any write violated the window, because the other end would send a reset (and that is tested by server.sent.overflow/ and client.sent.overflow).

So I think we should remove this test too.

NicoletaOita replied:
Done.

NicoletaOita commented:
@cmebarrow Thank you for the feedback, I've applied it: updated the readme, replaced nukleus:throttle "off" with nukleus:throttle "none", removed the tests. However, there is one I think you've missed, SimplexIT.shouldThrottleInitialOnlyClientSentDataTest. I've added it just to ask about the expected behavior and what exactly does the option "update" do. So if I have a long message such that the initial window is filled up and "update" is set to none, should it result in not receiving the rest of the message?

cmebarrow commented on README.md:
```diff
@@ -213,3 +213,8 @@ read aborted
+
+NOTE:
+The "nukleus:window" option specifies the number of bytes that will be advertised in the first Window frame sent back in response to a Begin
+The "nukleus:throttle" option determines how and whether writes will respect Window frames received from the other end. When set to "stream" or "message", window frames are honored, When set to "none" they are not.
```
the final When should be when (lower case).

NicoletaOita replied:
Ok.

cmebarrow commented:
@NicoletaOita Regarded your last question, yes I would expect it would result in not all the data being sent. So I would expect that test to fail with timeout. Is that not what you observe?

cmebarrow commented on src/test/scripts/org/reaktivity/k3po/nukleus/ext/simplex/server.sent.throttle.initial.only.update.none/client.rpt:
```diff
@@ -0,0 +1,25 @@
+connect "nukleus://server/streams/client"
+        option nukleus:route 1234L
+        option nukleus:transmission "simplex"
+        option nukleus:window 8192
```
remove this line (not relevant for simplex on client side)

NicoletaOita replied:
If I don't add this line, I get the following error: 
```
org.kaazing.k3po.junit.rules.SpecificationException: Internal error:nukleus window option missing
	at org.kaazing.k3po.junit.rules.ScriptRunner.call(ScriptRunner.java:135)
	at org.kaazing.k3po.junit.rules.ScriptRunner.call(ScriptRunner.java:42)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)


```

cmebarrow replied:
OK. I have filed https://github.com/reaktivity/k3po-nukleus-ext.java/issues/23 for this bug.

cmebarrow commented on src/test/scripts/org/reaktivity/k3po/nukleus/ext/simplex/server.sent.throttle.initial.only.update.none/server.rpt:
```diff
@@ -0,0 +1,26 @@
+accepted
+connected
+
+read "Hello, world"
```
This test should be failing with timeout but it's not (because the Travis build passed, as David pointed out). So we have a bug to fix. However this test should be written differently to verify the expected behavior, since relying on the timeout is not a good method (slows up the build and it's not precise). What I would suggest you do is as follows (instead of read "Hello, world"), combined with the changes I suggested in my comment on the other script:
```
read "H"
read notify FIRST_BYTE_RECEIVED
```
At this point the other script will send a second byte which should cause this script to automatically send a reset because it violates the total we know we gave the client (still only 1 because of the update:none option)

NicoletaOita replied:
Ok, I've updated it.

cmebarrow commented on src/test/scripts/org/reaktivity/k3po/nukleus/ext/simplex/server.sent.throttle.initial.only.update.none/client.rpt:
```diff
@@ -0,0 +1,25 @@
+connected
+
+write "Hello"
+write ", "
```
Please read my comment on the other script before reading this one. Where I suggest you do in this script is as follows:
```
write "H"
write await FIRST_BYTE_RECEIVED
write "e"
write aborted
```
Combined with the changes I suggested to the other script the test should pass. Given that it is currently passing I would expect that it will probably fail because you won't see the "write aborted". Once you have made the test changes and confirmed the test fails then I can work on fixing it so the test passes.

NicoletaOita replied:
I've updated the scripts, both client and server as you requested and the test fails. I've filed an issue for it: https://github.com/kaazing/tickets/issues/1206 and ignored the test in this PR until it will be fixed.

NicoletaOita commented:
@cmebarrow The test was passing as it was initially and I was expecting it to fail with timeout.
After changing the scripts as you said, the test is failing because no reset is sent (no"write aborted") and I would expect it to pass.
Please merge this PR if it's ok (I've applied all the feedback except for the window option which causes error if removed) and we'll track this issue [here](https://github.com/kaazing/tickets/issues/1206).

cmebarrow commented on src/test/java/org/reaktivity/k3po/nukleus/ext/SimplexIT.java:
```diff
@@ -176,6 +177,17 @@ public void shouldThrottleInitialOnlyClientSentData() throws Exception
         k3po.finish();
     }
 
+    @Ignore ("Please see issue #1206: https://github.com/kaazing/tickets/issues/1206")
```
Please replace this bug with one filed against reaktivity/k3po-nukleus-ext, as I requested in [the bug](https://github.com/kaazing/tickets/issues/1206), and update this link point to the new one.

NicoletaOita replied:
Done.

cmebarrow commented:
@jfallows or @dpwspoon Please do final review and merge

NicoletaOita commented:
@dpwspoon The requested changes left which appear [above](https://github.com/reaktivity/k3po-nukleus-ext.java/pull/20#pullrequestreview-64482322) don't apply anymore . I've removed that test according to Chris's feedback.
